### PR TITLE
refactor API and add Dashboard & API Server

### PR DESCRIPTION
**Pull request by sumory**

### CODE REVIEW

File: gateway/api.lua [hunk 1]
```diff
@@ -0,0 +1,95 @@
+local API = {}
+local table_insert = table.insert
+local ipairs = ipairs
+local type = type
+local tostring = tostring
+local cjson = require("cjson")
```
Comment (moonbingbing) [inline gateway/api.lua]: unused variables at line 2 to line 6

File: gateway/api.lua [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+local status = require("gateway.module.status")
+local summary = require("gateway.module.summary")
+
+
+API["/status"] = {
+    GET = function()
+        return function(req, res, next)
+            local resport = status.report()
```
Comment (moonbingbing) [inline gateway/api.lua:10]: if status.report() has 500 error, error.log can not log the error info and stack. 

```
local common = require "gateway.module.common"

xpcall(status.init, common.err_handle)
```

can fix this problem.

Comment (sumory) [inline gateway/api.lua:10]: the thirdparty framework `lor` would catch these errors.

Comment (sumory) [inline gateway/api.lua]: the thirdparty framework `lor` would catch these errors.

File: gateway/api.lua [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+            if resport then
+                res:json({
+                    success = true,
+                    msg = "",
+                    data = resport
+                }, true)
```
Comment (moonbingbing) [inline gateway/api.lua]: the `/status` interface is not compatible the NGINX Plus, by add one more json layer like `msg` and `data`

Comment (sumory) [inline gateway/api.lua]: the API exposed at port 9090 is mainly aimed at common usages for any other third party applications. so i would like to assume that it should have a normal response format just like the following:

```
{
    success: 0, //(true/false or other error code)
    msg: "description for this result",
    data: { // the real response data, such as `Nginx Plus status object`
    }
}
```

Comment (moonbingbing) [inline gateway/api.lua]: in my opinion, the `GET` API should directly return result. 
- If failed, return empty and standard http status code, like 4xx, 5xx. 
- if success, just return what they want without additional judge and parse.
  Maybe we can take a look at [twitter API](https://dev.twitter.com/rest/public) and [Uber API](https://developer.uber.com/docs/rides/api-overview).

BTW `datadog` support NGINX Plus's API format, if we use a different format, we will lose many OPS/APM  companies support.

Comment (sumory) [inline gateway/api.lua]: if we just want a response status like 4xx/5xx, that could be done with:

```
if resport then
    res:json({
        success = true,
        msg = "",
        data = resport
    }, true)
else
    res:status(500)
end
```

and i think the response description(JSON/text/html...) could always be helpful for the API protocol whether the `HTTP status code` is 200 or not.

as a uniform style for some scenes such as frontend usages(ajax API),  some response text(JSON) that stands for what happened may be a better choice.

any way i will refactor this part if this is the preference for Mio. 😃

File: api/router.lua [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+else
+    if api and type(api) == "table" then
+        for uri, api_methods in pairs(api) do
+            ngx.log(ngx.INFO, "load route, uri:", uri)
+            if type(api_methods) == "table" then
+                for method, func in pairs(api_methods) do
+                    local m = string_lower(method)
+
+                    if m == "get" then
+                         api_router:get(uri, func(store))
```
Comment (moonbingbing) [inline api/router.lua]: what's `store` means? it seems like a unused variable

File: api/router.lua [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+                    elseif m == "put" then
+                        api_router:put(uri, func(store))
+                    elseif m == "delete" then
+                        api_router:delete(uri, func(store))
+                    end
+                end
+            end
+        end
+    end
+end
```
Comment (moonbingbing) [inline api/router.lua:28]: there are too many `end` to understand the logic of the program. Maybe we can optimize the logic of this part

Comment (sumory) [inline api/router.lua:28]: done!

File: api/router.lua (hunk 4 above)
Comment (sumory) [inline api/router.lua]: :cry:  this part is copied from my another project without careful consideration.

**Comment by moonbingbing** [comment]
we need add more detail docs after this PR for understanding how to use.

**Pull request merged**